components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java

```diff
@@ -0,0 +1,141 @@
+                }
+            } else {
+                //remove existing files inside the directory
+                FileUtils.cleanDirectory(folder);
+            }
+
+            //get the zip file content
+            zipInputStream = new ZipInputStream(themeFile);
+            //get the zipped file list entry
+            ZipEntry zipEntry = zipInputStream.getNextEntry();
+
+            while (zipEntry != null) {
+
+                String fileName = zipEntry.getName();
+                APIUtil.validateFileName(fileName);
+                File newFile = new File(outputFolder + File.separator + fileName);
+                String canonicalizedNewFilePath = newFile.getCanonicalPath();
+                String canonicalizedDestinationPath = new File(outputFolder).getCanonicalPath();
+                if (!canonicalizedNewFilePath.startsWith(canonicalizedDestinationPath)) {
+                    handleException(
+                            "Attempt to upload invalid zip archive with file at " + fileName + ". File path is " +
+                                    "outside target directory");
+                }
+
+                if (zipEntry.isDirectory()) {
+                    if (!newFile.exists()) {
+                        boolean status = newFile.mkdir();
+                        if (!status) {
+                            handleException("Error while creating " + newFile.getName() + " directory");
+                        }
+                    }
+                } else {
+                    String ext = FilenameUtils.getExtension(zipEntry.getName());
+                    if (EXTENSION_WHITELIST.contains(ext)) {
+                        //create all non exists folders
+                        //else you will hit FileNotFoundException for compressed folder
+                        new File(newFile.getParent()).mkdirs();
+                        FileOutputStream fileOutputStream = new FileOutputStream(newFile);
+
+                        int len;
+                        while ((len = zipInputStream.read(buffer)) > 0) {
+                            fileOutputStream.write(buffer, 0, len);
+                        }
+
+                        fileOutputStream.close();
+                    } else {
+                        handleException(
+                                "Unsupported file is uploaded with tenant theme by " + tenantDomain + " : file name : "
+                                        + zipEntry.getName());
+                    }
+
+                }
+                zipEntry = zipInputStream.getNextEntry();
+            }
+
+            zipInputStream.closeEntry();
+            zipInputStream.close();
+
+        } catch (IOException ex) {
+            handleException("Failed to deploy tenant theme", ex);
+        } finally {
+            IOUtils.closeQuietly(zipInputStream);
+            IOUtils.closeQuietly(themeFile);
+        }
+    }
+
+    private static void handleException(String msg) throws APIManagementException {
+
```

Comment:
check whether we can reuse org.wso2.carbon.apimgt.impl.utils.APIUtil.handleException;

Reply:
Fixed with https://github.com/wso2/carbon-apimgt/pull/8595/commits/6ae0bc73c2aefce667fc3908529e7750aaa25d97